### PR TITLE
Fix for issue 984 and 992

### LINKED ISSUE #992: Multi-level interfaces are broken
# Brief Description

If an interface implements another interface, any class implementing the second interface won't have any interface methods automatically generated by Umple.

Also, if an interface has the above problem, and class extending it will also have the problem (this is a separate issue).

## Minimum Steps to Reproduce

Try generating Java from InterfaceTemplateTest.ump using Umple online. Any class that implements ISecondChild has this problem.

## Expected Result

All interface methods should be generated

## Actual Result

None are being generated

### PR DESCRIPTION
**Pull request by AdamBJ**
## PLEASE READ THIS BEFORE MERGING THE PULL REQUEST
This PR populates 15 previously empty expected output files used in the JUnit template tests. These files are associated primarily with state machines. Before merging this PR, someone needs to verify that the output I've added is correct given the existing input `.ump` files. I've done my best to do this, but I'm not that familiar with state machines.

For example, someone needs to look at `stateMachineWithStringComparisonGuard.ump` and verify that the output added to `stateMachineWithStringComparisonGuard.php.txt` is the output one would expect if Umple was run with `stateMachineWithStringComparisonGuard.ump`. There are around 15 such verification steps required. I think this might be a good task for @jblang94, it might be best to create a new issue to address this task. Please see [this commit](https://github.com/umple/umple/commit/46ec5e94c8aa5ed015f9c99ccfc3a5301f29e6ce) for a complete list of the expected output files that were populated.

## Description

This PR adds a check to `assertFileContent` to make sure that the expected output file in not empty. Adding this check revealed 15 expected output files that were empty. I have run Umple on the input files associated with these empty expected output files and populated them with the results. Verification is required before this PR can be merged with master.

This PR also fixes [issue 992](https://github.com/umple/umple/issues/992) (see [this commit](https://github.com/umple/umple/commit/c1b86996652d57684b02df560b7efb88beb0438e)).

## Tests

This PR populated all the empty expected output files uncovered by the issue 984 fix, and deleted one duplicate test. It also modifies an existing test case and adds more test cases to ensure that the fix introduced to address issue 992 is correct.

`Closes #992, #984`

### CODE REVIEW

**Comment by Nava2** [comment]
Failed tests are from your changes

```
[junit] Running cruise.umple.implementation.java.JavaInterfaceTemplateTest
[junit] Tests run: 7, Failures: 1, Errors: 0, Time elapsed: 0.056 sec
```

Though, is this caused because the test code itself was changed? Thus, meaning it will fail when run against old versions?

**Comment by AdamBJ** [comment]
@Nava2 Yeah, I added some bit more code to a couple of the .ump files. I guess it would be better to just split it into a second file, I'll do that instead.

**Comment by AdamBJ** [comment]
Though I don't understand why it would be green for me locally but fail the build in travis?

File: cruise.umple/src/UmpleInternalParser_CodeClass.ump [hunk 1]
```diff
@@ -1602,20 +1602,20 @@ private Boolean checkIsDistributed(UmpleInterface uInterface)
...
-		          aMethod.setSource(Method.Source.fAuto);
-		          uClass.addMethod(aMethod);
-		        }
-		      }
-		    }
+		if (pi.hasMethods())
+		{
+		  for (Method aMethod : pi.getMethods())
+		  {
+			boolean shouldAddMethod = verifyIfMethodIsConstructorOrGetSet(uClass, aMethod, true);
```
Comment (AdamBJ) [inline cruise.umple/src/UmpleInternalParser_CodeClass.ump]: This looks fine in my local editor (Notepad++). Working on a fix.

**Comment by AdamBJ** [comment]
@Nava2 Thanks for catching that

File: cruise.umple/src/UmpleInternalParser_CodeClass.ump (hunk 1 above)
Comment (Nava2) [reply cruise.umple/src/UmpleInternalParser_CodeClass.ump]: Check if your editor is set to tabs vs. spaces, you should use spaces and 2 per tab! 👍

**Comment by vahdat-ab** [comment]
In the following example, Umple doesn't generate default implementation to the method in the interface `I`.
```
interface I{
  void test();
}
interface I2{
  isA I;
  void test2();
}
interface I3{
  isA I2;
  void test3();
}
class A{
  isA I3;
}
```

BTW, Please always make sure to request a separate PR for each issue. Never mix them, except if it's about typos in  wiki

**Comment by vahdat-ab** [comment]
@opeyemiAdesina please check the test cases related to your work.

**Comment by AdamBJ** [comment]
I'll have a look at that issue, thanks. I was originally going to make this PR just for 994, but 994 can't be resolved until the interface hierarchy stuff in 992 is dealt with (one of the tests that was blank was testing the interface hierarchy stuff).

**Comment by vahdat-ab** [comment]
I checked the code related to state machines in Java and it looks ok.
Regarding the empty files, please check the history of them. Looks like they were ok until someone completely removed them. Check [this ](https://github.com/umple/umple/commits/c1b86996652d57684b02df560b7efb88beb0438e/cruise.umple/test/cruise/umple/implementation/java/InterfaceTemplateTest_Methods.java.txt) out

**Comment by AdamBJ** [comment]
Hm, that's weird. I guess someone went a little crazy with the find/replace, haha. Did you find that for any other files aside from the stuff testing the interface hierarchy? I did a check of some of the other blank files and they seem blank from the beginning.

**Comment by AdamBJ** [comment]
I'll use the old tests for the interface hierarchy stuff if there are any discrepancies between the the test I added and the old tests.

**Comment by vahdat-ab** [comment]
I didn't see. Please explore the necessary ones. Good luck :-)

**Comment by AdamBJ** [comment]
@vahdat-ab I've pushed a fix for the interface hierarchy code and added the test case for it you provided to the JUnit tests. I've also checked the expected output files for the interface tests, there aren't any significant differences between the output I added and the original stuff from 2011. All the other expected output files started empty.

Whenever the remaining expected output files have been verified I think this PR can be merged.

**Comment by vahdat-ab** [comment]
@opeyemiAdesina I wonder if you will be able to have a look at the modified examples related to your work in this PR soon?

**Comment by vahdat-ab** [comment]
Thanks @opeyemiAdesina
@AdamBJ Please update your repository to see new modification done by @opeyemiAdesina. If everything is fine, let me know and I'll merge it.

**Comment by vahdat-ab** [comment]
@AdamBJ pay attention that there is a conflict because Opeyemi updated his files.

**Comment by AdamBJ** [comment]
@opeyemiAdesina I had a look at the file you added (`ExampleFile.alloy.txt`), seems like it's the same as the one I originally deleted. The reason I got rid of it is because these two tests:

```
 @Test //@Ignore
	  public void ExampleFile()
	  {
		assertUmpleTemplateFor("alloy/ExampleFile.ump","alloy/ExampleFile.alloy.txt");
		Assert.assertEquals(true, (new File(pathToInput + "/alloy/ExampleFile.als")).exists());
	  }
	  
	  @Test //@Ignore
	  public void BankingSystem()
	  {
		assertUmpleTemplateFor("alloy/BankingSystem.ump","alloy/BankingSystem.alloy.txt");
		Assert.assertEquals(true, (new File(pathToInput + "/alloy/BankingSystem.als")).exists());
	  }
```

Seem to be exactly the same. BankingSystem.ump is identical to ExampleFile.ump, as are ExampleFile.alloy.txt and BankingSystem.alloy.txt. It seems like the tests are testing the exact same thing, so I got rid of one of them. If that was a mistake let me know.

**Comment by vahdat-ab** [comment]
@opeyemiAdesina I wonder if you can have a look at this PR and if it's ok then I can merge it.

**Comment by AdamBJ** [comment]
@vahdat-ab @opeyemiAdesina We just need to decide whether to keep the `ExampleFile()` test or not. I don't think we should, since it's identical to the `BankingSystem()` test below it.

**Comment by opeyemiAdesina** [comment]
I have always used the file for testing new features before creating
additional tests. However, since I don't intend to make any changes at the
moment you may remove it.

**Comment by vahdat-ab** [comment]
@AdamBJ please resolve the conflict shown in your PR and then I'll merge it,
thanks @opeyemiAdesina

**Comment by AdamBJ** [comment]
@vahdat-ab I made a mistake when I rebased 984. Working on a fix so that the commit history isn't so messy

**Comment by AdamBJ** [comment]
I can't figure out a way to clean up the commit history without creating a new branch. I'm going to close this PR and open a new one from the new branch with clean commit history.

**Comment by Nava2** [comment]
@AdamBJ you can do a `reset --hard COMMITISH` then `rebase ORIGINAL-COMMITISH`. Make sure you keep track of the commits you care about, or just do a merge. In this context, the merge is likely worth a commit.

**Comment by AdamBJ** [comment]
@Nava2 How can I push those changes to the remote repo? I was able to get my local repo reverted doing something like you suggest, but when I tried to push to the remote it said I couldn't since my local branch was behind the remote branch. At this point I'm afraid I've made such a mess of things that I just want to amputate the mess and start again.

**Comment by AdamBJ** [comment]
I've never used COMMITISH before, I'm having some trouble though:

```
$ git reset --hard COMMITISH
fatal: ambiguous argument 'COMMITISH': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
```

Do I need to run this command for every commit I don't want in the history?

**Comment by Nava2** [comment]
> @Nava2 How can I push those changes to the remote repo? I was able to get my local repo reverted doing something like you suggest, but when I tried to push to the remote it said I couldn't since my local branch was behind the remote branch. At this point I'm afraid I've made such a mess of things that I just want to amputate the mess and start again.

In this case, teh issue is that you would have to "force" push. You've rewritten history, thus you'd have to force the remote repository accept it. 

> I've never used COMMITISH before, I'm having some trouble though:

A "commitish" is just either: 1) commit, 2) branch, 3) remote branch, 4) tag, etc.. Anything you can use to reference a commit tree.

**Pull request closed without merging**